src/main/java/org/wildfly/security/auth/server/RealmIdentity.java

```diff
@@ -266,4 +284,28 @@ public boolean exists() throws RealmUnavailableException {
             return false;
         }
     };
+
+    /**
+     * Represents a key that is specific to the realm and used to uniquely identify a {@link RealmIdentity} instance.
+     */
+    interface Key {
+
+        Key EMPTY = (Key) () -> null;
+
+        /**
+         * Returns the value of this key.
+         *
+         * @return the value of this key
+         */
+        Object getValue();
```

Comment:
Originally we added it because of the weird combinations of name, principal, and evidence.  I was hung up on the idea that Principal should pertain only to SI, but really that's a pointless restriction.  A Principal is just an identifier.  Between Principal and Evidence I think we cover everything possible: the difference between the two is whether the information to use to locate the identity is public (Principal) or is either private or otherwise opaque to the domain (Evidence).  Since everything can fall into one of these two categories I think there is no reason for the extra complexity of IL.  That was my reasoning.